clients/pkg/promtail/client/client_writeto.go

```diff
@@ -0,0 +1,90 @@
+package client
+
+import (
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/tsdb/record"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/ingester/wal"
+	"github.com/grafana/loki/pkg/util"
+)
+
+type seriesCacheEntry struct {
```

Comment:
actually, the reason we have the labels cache and `lastSeen` map separate in prometheus is to avoid the lock contention possibility

I would tend to think the memory overhead of the 2nd map is worth it to split the two and avoid lock contention; this is especially true since at the moment we write a series record every time we see a log line since we're not tracking the series we already know about when writing to the WAL. Happy to have you prove otherwise if you would like to keep this struct and a single map. I think we could write benchmark that has a few go routines all writing lots of log entries and see which is able to write faster.

Reply:
I read over the queue manager implementation, and I like the idea of keeping entries across two maps, one with the labelsets, and once with the last seen segment. Regardless, I think it won't have any effect at this point of the implementation, since the main contention point would occurs when: 
- the watcher is running and appending entries/storing series
- a checkpoint is detected, and the [`UpdateSeriesSegment`](https://github.com/prometheus/prometheus/blob/8f3b2fc9aeaf1bd3f839e820e7f9ffc571006e89/storage/remote/queue_manager.go#L844) in the `WriteTo` is called, avoiding a possible contention if both were a single entry

And the Promtail WAL doesn't use checkpoints for the time being. Since the overhead is almost none, It'd be awesome to keep it so in the future we avoid possible contentions.


```
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/clients/pkg/promtail/client
                                                                     │ pr_single_map.txt │           pr_dual_map.txt            │
                                                                     │      sec/op       │   sec/op     vs base                 │
ClientWriteTo/num_writers_5,_with_1000_lines_written_in_total-10             1.433 ±  8%   1.362 ± 14%       ~ (p=0.710 n=10)
ClientWriteTo/num_writers_10,_with_1000_lines_written_in_total-10            1.427 ± 12%   1.485 ±  4%       ~ (p=0.201 n=10)
ClientWriteTo/num_writers_5,_with_10000_lines_written_in_total-10            1.970 ± 14%   2.007 ± 20%       ~ (p=0.503 n=10)
ClientWriteTo/num_writers_10,_with_10000_lines_written_in_total-10           1.762 ±  5%   1.722 ±  3%       ~ (p=0.541 n=10+8)
ClientWriteTo/num_writers_5,_with_100000_lines_written_in_total-10           6.628 ±  5%   6.892 ±  2%  +3.97% (p=0.037 n=10+8)
ClientWriteTo/num_writers_10,_with_100000_lines_written_in_total-10          4.033 ±  5%   4.000 ±  3%  -0.82% (p=0.037 n=10+8)
ClientWriteTo/num_writers_10,_with_1000000_lines_written_in_total-10         27.12 ±  0%   27.03 ±  1%       ~ (p=0.201 n=10)
geomean                                                                      3.390         3.396        +0.16%

                                                                     │ pr_single_map.txt │            pr_dual_map.txt             │
                                                                     │       B/op        │     B/op      vs base                  │
ClientWriteTo/num_writers_5,_with_1000_lines_written_in_total-10            35.39Ki ± 6%   35.30Ki ± 3%        ~ (p=0.503 n=10)
ClientWriteTo/num_writers_10,_with_1000_lines_written_in_total-10           43.96Ki ± 4%   48.91Ki ± 5%  +11.27% (p=0.000 n=10)
ClientWriteTo/num_writers_5,_with_10000_lines_written_in_total-10           312.3Ki ± 0%   312.8Ki ± 1%        ~ (p=0.056 n=10)
ClientWriteTo/num_writers_10,_with_10000_lines_written_in_total-10          316.2Ki ± 1%   317.1Ki ± 0%        ~ (p=0.090 n=10+8)
ClientWriteTo/num_writers_5,_with_100000_lines_written_in_total-10          3.055Mi ± 0%   3.055Mi ± 0%        ~ (p=0.794 n=10+8)
ClientWriteTo/num_writers_10,_with_100000_lines_written_in_total-10         3.053Mi ± 0%   3.051Mi ± 0%        ~ (p=0.334 n=10+8)
ClientWriteTo/num_writers_10,_with_1000000_lines_written_in_total-10        30.54Mi ± 0%   30.54Mi ± 0%   -0.01% (p=0.010 n=10)
geomean                                                                     646.1Ki        656.2Ki        +1.56%

                                                                     │ pr_single_map.txt │           pr_dual_map.txt            │
                                                                     │     allocs/op     │  allocs/op   vs base                 │
ClientWriteTo/num_writers_5,_with_1000_lines_written_in_total-10             1.169k ± 1%   1.163k ± 1%       ~ (p=0.909 n=10)
ClientWriteTo/num_writers_10,_with_1000_lines_written_in_total-10            1.236k ± 0%   1.248k ± 1%  +0.97% (p=0.000 n=10)
ClientWriteTo/num_writers_5,_with_10000_lines_written_in_total-10            18.88k ± 0%   18.88k ± 0%       ~ (p=0.392 n=10)
ClientWriteTo/num_writers_10,_with_10000_lines_written_in_total-10           17.69k ± 0%   17.68k ± 0%       ~ (p=0.405 n=10+8)
ClientWriteTo/num_writers_5,_with_100000_lines_written_in_total-10           198.9k ± 0%   198.9k ± 0%       ~ (p=0.636 n=10+8)
ClientWriteTo/num_writers_10,_with_100000_lines_written_in_total-10          197.7k ± 0%   197.7k ± 0%       ~ (p=0.117 n=10+8)
ClientWriteTo/num_writers_10,_with_1000000_lines_written_in_total-10         1.998M ± 0%   1.998M ± 0%  -0.00% (p=0.010 n=10)
geomean                                                                      32.45k        32.47k       +0.06%
```

Reply:
You're right, I was mis-remembering. The watcher is single threaded and can't concurrently append samples to `writeTo`and also call `seriesReset`.

Reply:
> The watcher is single threaded
Yes! But `AppendSeries` and `SeriesReset` can be called concurrently since one is called by the watcher main thread, when appending series, and the other by the "whomever" is polling for deleted segments: this can be the current impl, which is another thread that lives in the watcher, or the notifier you mention.
Regardless, both of those concurrent call can't be optimized that much since they lock the main cache map.